mwdb/resources/group.py

```diff
@@ -262,21 +262,30 @@ def put(self, name, login):
 
         user_login_obj = load_schema({"login": login}, UserLoginSchemaBase())
 
-        member = db.session.query(User).filter(User.login == login).first()
-        if member is None:
-            raise NotFound("No such user")
-
-        if member.pending:
-            raise Forbidden("User is pending and need to be accepted first")
-
         group = db.session.query(Group).options(joinedload(Group.users)).filter(Group.name == name).first()
+        if not (g.auth_user.has_rights(Capabilities.manage_users) or g.auth_user.is_group_admin(group.id)):
+            raise Forbidden("You are not permitted to manage this group")
+
         if group is None:
             raise NotFound("No such group")
 
         if group.immutable:
             raise Forbidden("Adding members to private or public group is not allowed")
 
-        group.users.append(member)
+        member = db.session.query(User).filter(User.login == login).first()
+        if member is None:
+            raise NotFound("No such user")
+
+        if member.pending:
+            raise Forbidden("User is pending and need to be accepted first")
+
+        if group in member.groups and g.auth_user.has_rights(Capabilities.manage_users):
+            if member.is_group_admin(group.id):
+                member.set_group_admin(group.id, False)
+            else:
+                member.set_group_admin(group.id, True)
+        else:
+            group.users.append(member)
```

Comment:
That's not correct PUT implementation, an effect should not rely on the state of object.

1. `append` will work even if member is already added. It's based on foreign keys so it works as set, not as list.
2. Sometimes we may want to add new admin member via REST API (both set as admin and added as member). In addition; that change should be made within the same transaction.

The 2. may be complicated without changes in model:
- Member must be regular class-based table, with one-to-many relationships on each side.
- To keep User and Group behave like many-to-many relationship, we need to bind them using association proxies (https://docs.sqlalchemy.org/en/13/orm/extensions/associationproxy.html)

I can help you with that change.